File: openmole/misc/org.openmole.misc.hashservice/src/main/java/org/openmole/misc/hashservice/SHA1Hash.java
/*
 *  Copyright (C) 2010 reuillon
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the Affero GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.openmole.misc.hashservice;

import java.io.UnsupportedEncodingException;
import java.util.Arrays;
import org.openmole.commons.exception.FatalError;
import org.openmole.commons.tools.io.IHash;

//Should stay in the public part for serialization purposes
public class SHA1Hash implements IHash {

    byte[] content;
    static final byte[] HEX_CHAR_TABLE = {
        (byte) '0', (byte) '1', (byte) '2', (byte) '3',
        (byte) '4', (byte) '5', (byte) '6', (byte) '7',
        (byte) '8', (byte) '9', (byte) 'a', (byte) 'b',
        (byte) 'c', (byte) 'd', (byte) 'e', (byte) 'f'
    };

    public static String getHexString(byte[] raw) throws UnsupportedEncodingException {
        byte[] hex = new byte[2 * raw.length];
        int index = 0;

        for (byte b : raw) {
            int v = b & 0xFF;
            hex[index++] = HEX_CHAR_TABLE[v >>> 4];
            hex[index++] = HEX_CHAR_TABLE[v & 0xF];
        }
        return new String(hex, "ASCII");
    }

    public SHA1Hash(byte[] content) {
        super();
        this.content = content;
    }

    @Override
    public String toString() {
        try {
            return getHexString(content);
        } catch (UnsupportedEncodingException e) {
            throw new FatalError(e);
        }
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + Arrays.hashCode(content);
        return result;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        SHA1Hash other = (SHA1Hash) obj;
        if (!Arrays.equals(content, other.content)) {
            return false;
        }
        return true;
    }

    @Override
    public int compareTo(IHash o) {
        return toString().compareTo(o.toString());
    }
}

File: openmole/misc/org.openmole.misc.hashservice/src/main/scala/org/openmole/misc/hashservice/SHA1Hash.scala
/*
 * Copyright (C) 2010 reuillon
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.openmole.misc.hashservice


import org.openmole.misc.tools.service.IHash

class SHA1Hash(val content: Array[Byte]) extends IHash {

  override def toString: String = IHash.hexString(content)

  override def hashCode: Int = content.deep.hashCode

  override def equals(obj: Any): Boolean = {
    if (obj == null) return false
    if (getClass != obj.asInstanceOf[AnyRef].getClass) return false
    val other = obj.asInstanceOf[SHA1Hash]
    content.deep == other.content.deep
  }

}
